### PR TITLE
Base Immutable class

### PR DESCRIPTION
**Pull request by vogievetsky**
The idea is that instead of just implementing the provided interface immutable classes can extend a class and have that class do a lot of the work.

This might need a little bit of iteration before we get it right.

### CODE REVIEW

File: src/baseImmutable.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    for (var property of properties) {
+      var propertyName = property.name;
+      var equal = property.equal || generalEqual;
+      if (!equal((this as any)[propertyName], (other as any)[propertyName])) return false;
+    }
+
+    return true;
+  }
+
+  public get(propName: string): any {
```
Comment (lorem--ipsum) [inline src/baseImmutable.ts]: This class could also use a `public set(propName: string, propValue: any): BaseImmutable` method, right ?

File: src/baseImmutable.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  constructor(value: ValueType) {
+    var properties = this.getProperties();
+    for (var property of properties) {
+      var propertyName = property.name;
+      (this as any)[propertyName] = (value as any)[propertyName];
+    }
+  }
+
+  public getProperties(): Property[] {
+    return (this.constructor as any).PROPERTIES;
```
Comment (lorem--ipsum) [inline src/baseImmutable.ts]: I think this should throw an error if `PROPERTIES` doesn't exist.

File: src/baseImmutable.ts (hunk 1 above)
Comment (vogievetsky) [inline src/baseImmutable.ts]: just added (`set`), see below

Comment (lorem--ipsum) [inline src/baseImmutable.ts]: Awesome ! And yeah, here's a good place to use the polymorphic return type !

**Comment by vogievetsky** [comment]
goody

**Pull request merged**